### PR TITLE
refactor topology and constructor of Graph

### PR DESCRIPTION
**Pull request by houpc**
Refactor: in the new convention, topology includes the contractions between ghost operators and external operators, forming fake legs.
Modify the relevant tests.

### CODE REVIEW

File: src/computational_graph/graph.jl [hunk 1]
```diff
@@ -136,20 +136,57 @@ is_external(g::Graph, i::Int) = i in g.external
...
+    Return all the external operators with real legs (::Vector{OperatorProduct}).
+"""
+real_extV(g::Graph) = filter(!isghost, external_legs(g))
+
+"""
+    function fake_extV(g::Graph)
+
+    Return all the external operators with fake legs (::Vector{OperatorProduct}).
+"""
+function fake_extV(g::Graph)
```
Comment (kunyuan) [inline src/computational_graph/graph.jl:179]: fake_extV  --> ghost_external

File: src/computational_graph/graph.jl [hunk 2]
```diff
@@ -136,20 +136,57 @@ is_external(g::Graph, i::Int) = i in g.external
...
     Return all external vertices (::Vector{QuantumOperators}) of Graph `g`.
 """
 external(g::Graph) = OperatorProduct(g.vertices)[g.external]
 
+"""
+    function external_legs(g::Graph)
+
+    Return all the external legs, including real and ghost external vertices (::Vector{OperatorProduct}).
+"""
+external_legs(g::Graph) = OperatorProduct(g.vertices)[eachindex(g.external)]
```
Comment (kunyuan) [inline src/computational_graph/graph.jl:158]: external(g::Graph) = g.vertices[g.external]

File: src/computational_graph/graph.jl [hunk 3]
```diff
@@ -136,20 +136,57 @@ is_external(g::Graph, i::Int) = i in g.external
 """
 is_internal(g::Graph, i::Int) = (i in g.external) == false
 
+"""
+    function is_ghost(g::Graph, i::Int) 
+
+    Check if `i::Int` in the ghost operator's indices of Graph `g`.
+"""
+is_ghost(g::Graph, i::Int) = isghost(OperatorProduct(g.vertices)[i])
```
Comment (kunyuan) [inline src/computational_graph/graph.jl:144]: is_ghost and isghost should be unified

**Review by kunyuan — APPROVED** [review]

File: test/computational_graph.jl [hunk 4]
```diff
@@ -117,83 +115,106 @@ end
...
-        g8 = feynman_diagram(V8, [[1, 3], [2, 6], [4, 14], [5, 13], [7, 10], [8, 11]], external=[1, 9, 11, 12])
+        V8 = [𝑓⁻(1, true), 𝑓⁺(2), 𝑓⁻(12), 𝑓⁺(16, true), 𝑓⁺(3)𝑓⁻(4)𝜙(5), 𝑓⁺(6)𝑓⁻(7)𝜙(8), 𝑓⁺(9)𝑓⁻(10)𝜙(11), 𝑓⁺(13)𝑓⁻(14)𝜙(15)]
+        g8 = feynman_diagram(V8, [[2, 6], [5, 9], [7, 16], [8, 15], [10, 13], [11, 3], [12, 4], [14, 1]], external=[2, 12, 3, 14])
         @test g8.factor == -1
-        gg8 = feynman_diagram(V8, [[1, 3], [2, 6], [4, 14], [5, 13], [7, 10], [8, 11]], external=[9, 1, 11, 12])
-        @test gg8.factor == -1
+        @test external(g8) == [𝑓⁺(2), 𝑓⁻(10), 𝑓⁻(12), 𝑓⁺(13)]
+        @test external_with_ghost(g8) == [𝑓⁻(1, true), 𝑓⁺(2), 𝑓⁻(12), 𝑓⁺(16, true)]
+        standardize_order!(g8)
+        @test external_with_ghost(g8) == [𝑓⁺(2), 𝑓⁺(16, true), 𝑓⁻(12), 𝑓⁻(1, true)]
```
Comment (dcerkoney) [inline test/computational_graph.jl:179]: Why is external(g8) ⊈ external_with_ghost(g8)? I wonder if it is a bug in `external_with_ghost`.

File: src/computational_graph/graph.jl [hunk 5]
```diff
@@ -150,6 +150,20 @@ external(g::Graph) = OperatorProduct(g.vertices)[g.external]
...
+    Return all physical external vertices (::Vector{OperatorProduct}) of Graph `g`.
+"""
+external(g::Graph) = OperatorProduct.(OperatorProduct(g.vertices)[g.external])
+
+"""
+    function external_with_ghost(g::Graph)
+
+    Return all the external vertices (::Vector{OperatorProduct}), including real legs and ghost.
+"""
+external_with_ghost(g::Graph) = OperatorProduct.(OperatorProduct(g.vertices)[eachindex(g.external)])
```
Comment (dcerkoney) [inline src/computational_graph/graph.jl:165]: I would prefer to split this function into `ghosts(g::Graph)`, and then `external_with_ghost(g::Graph) = union(ghosts(g), external(g))`.

File: test/computational_graph.jl (hunk 4 above)
Comment (dcerkoney) [reply test/computational_graph.jl:179]: Kun explained the conceptual misunderstanding to me; it is not a bug.

**Review by dcerkoney — APPROVED** [review]

**Pull request merged**